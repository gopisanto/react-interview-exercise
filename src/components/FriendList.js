import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styles from './FriendList.css';
import FriendListItem from './FriendListItem';
import Pagination from './Pagination';

const NO_OF_ITEMS_PER_PAGE = 2;
const NO_OF_PAGE_LINKS_TO_DISPLAY = 4;

export class FriendList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      activePage: 1,
    }

    this.handlePageClick = this.handlePageClick.bind(this);
  }

  handlePageClick(activePage) {
    this.setState({activePage});
  }

  getPaginatedFriends = () => {
    const {activePage} = this.state;
    const {friends} = this.props;
    const startIndex = activePage === 1 ? 0 : (activePage - 1) * NO_OF_ITEMS_PER_PAGE;
    const endIndex = startIndex + NO_OF_ITEMS_PER_PAGE - 1;

    return {paginatedFriends: friends.slice(startIndex, endIndex + 1), startIndex};
  }

  render () {    
    const {paginatedFriends, startIndex} = this.getPaginatedFriends();

    return (
      <div>
        <ul className={styles.friendList}>
          {
            paginatedFriends.map((friend, index) => {
              return (
                <FriendListItem
                  key={index}
                  id={startIndex + index}
                  name={friend.name}
                  starred={friend.starred}
                  gender={friend.gender}
                  {...this.props.actions} />
              );
            })
          }
        </ul>
        <Pagination
          noOfItems={this.props.friends.length}
          noOfItemsPerPage={NO_OF_ITEMS_PER_PAGE}
          noOfPageLinksToDisplay={NO_OF_PAGE_LINKS_TO_DISPLAY}
          onPageChange={this.handlePageClick}
        />
      </div>
    );
  }

}

FriendList.propTypes = {
  friends: PropTypes.array.isRequired,
  actions: PropTypes.object.isRequired
};

export default FriendList;