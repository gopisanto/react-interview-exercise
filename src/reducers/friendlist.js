import * as types from '../constants/ActionTypes';

export const initialState = {
  friendsById: [
    {
      name: 'Theodore Roosevelt',
      starred: true,
      gender: 'female'
    },
    {
      name: 'Abraham Lincoln',
      starred: false,
      gender: 'male'
    },
    {
      name: 'George Washington',
      starred: false,
      gender: 'male'
    }
  ]
};

export default function friends(state = initialState, action) {
  switch (action.type) {
    case types.ADD_FRIEND:
      const { name, gender } = action.payload;
      return {
        ...state,
        friendsById: [
          ...state.friendsById,
          {
            name,
            gender,
            starred: false
          }
        ],
      };
    case types.DELETE_FRIEND:
      return {
        ...state,
        friendsById: state.friendsById.filter((item, index) => index !== action.id)
      };
    case types.STAR_FRIEND:
      let friends = [...state.friendsById];
      let friend = friends.find((item, index) => index === action.id);
      friend.starred = !friend.starred;
      return {
        ...state,
        friendsById: friends
      };

    default:
      return state;
  }
}