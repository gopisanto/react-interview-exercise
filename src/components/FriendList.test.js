import React from 'react';
import renderer from 'react-test-renderer';
import {configure, shallow, mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import sinon from 'sinon';
import {FriendList} from './FriendList';
import {initialState} from '../reducers/friendlist';

configure({ adapter: new Adapter() });

describe('<FriendList />', () => {
    it('should render always same', () => {
        const friends = initialState.friendsById;
        const actions = {
            starFriend: () => null,
            deleteFriend: () => null
        };
        const wrapper = renderer.create(<FriendList friends={friends} actions={actions}  />).toJSON();

        expect(wrapper).toMatchSnapshot();
    });
    it('should handle page change', () => {
        const friends = initialState.friendsById;
        const actions = {
            starFriend: () => null,
            deleteFriend: () => null
        };
        const spy = sinon.spy(FriendList.prototype, 'handlePageClick');
        const wrapper = mount(<FriendList friends={friends} actions={actions}  />);
        wrapper.update();
        wrapper.find('span').findWhere(node => node.text() == '2').simulate('click');
        expect(spy.callCount).toBe(1);
        expect(spy.withArgs(2).calledOnce);
    });
});