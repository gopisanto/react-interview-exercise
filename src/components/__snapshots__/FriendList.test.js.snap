exports[`<FriendList /> should render always same 1`] = `
<div>
  <ul
    className={undefined}>
    <li
      className={undefined}>
      <div
        className={undefined}>
        <i
          aria-hidden="true"
          className="fa fa-2x fa-female" />
        <div
          className={undefined}>
          <div>
            <span>
              Theodore Roosevelt
            </span>
          </div>
          <div>
            <small>
              xx friends in common
            </small>
          </div>
        </div>
      </div>
      <div
        className={undefined}>
        <button
          className="btn btn-default undefined"
          onClick={[Function]}>
          <i
            className="fa fa-star" />
        </button>
        <button
          className="btn btn-default undefined"
          onClick={[Function]}>
          <i
            className="fa fa-trash" />
        </button>
      </div>
    </li>
    <li
      className={undefined}>
      <div
        className={undefined}>
        <i
          aria-hidden="true"
          className="fa fa-2x fa-male" />
        <div
          className={undefined}>
          <div>
            <span>
              Abraham Lincoln
            </span>
          </div>
          <div>
            <small>
              xx friends in common
            </small>
          </div>
        </div>
      </div>
      <div
        className={undefined}>
        <button
          className="btn btn-default undefined"
          onClick={[Function]}>
          <i
            className="fa fa-star-o" />
        </button>
        <button
          className="btn btn-default undefined"
          onClick={[Function]}>
          <i
            className="fa fa-trash" />
        </button>
      </div>
    </li>
  </ul>
  <div>
    <div
      className={undefined}>
      <span
        className="undefined"
        disabled={true}
        onClick={[Function]}>
        <<
      </span>
      <span
        className="undefined"
        disabled={true}
        onClick={[Function]}>
        <
      </span>
      <span
        className="undefined"
        onClick={[Function]}>
        1
      </span>
      <span
        className=""
        onClick={[Function]}>
        2
      </span>
      <span
        className=""
        disabled={false}
        onClick={[Function]}>
        >
      </span>
      <span
        className=""
        disabled={false}
        onClick={[Function]}>
        >>
      </span>
      <span
        className={undefined}>
        1 of 2
      </span>
    </div>
  </div>
</div>
`;
